feat: Complete user signup

minzzzy opened the pull request:
Resolves: #16

minzzzy committed:
feat: Add sign-up UserRestController with unit test

- Add column mapping in user entity
- Add user repository
- Add sign-up dto
- Add user service with test
- Add passwordEncoder bean in config

minzzzy committed:
Merge branch 'develop' into feature/signup

minzzzy committed:
feat: Add signupDto validation test

- Add valid annotation in controller
- Refactor to valid signup dto

minzzzy committed:
refactor: Change signup dto max name size

- Remove size annotation password in user
- Replace signup dto max name from 45 to 40

minzzzy committed:
Merge branch 'develop' into feature/signup

minzzzy committed:
refactor: Resolve conflicts signup with login logic

minzzzy committed:
refactor: Add response status in conflictException

minzzzy committed:
refactor: Redirect home when login

minzzzy committed:
refactor: Import javascript and change password regular expression

minzzzy committed:
feat: Add signup front-end

minzzzy committed:
fix: Fix wrong package path

minzzzy committed:
refactor: Add login logic

minzzzy committed:
fix: Remove user service layer for fixing directory name

Lahavas commented on src/main/java/com/woowahan/techcamp/recipehub/user/domain/UserRepository.java:
```diff
@@ -5,6 +5,6 @@
 
 public interface UserRepository extends JpaRepository<User, String> {
-
+    Boolean existsByEmail(String email);
```
따로 이유가 없다면 Boolean Class보단 boolean Type으로 Return하는게 좋을듯합니다

minzzzy replied:
이해했습니다. 수정할께요

Lahavas commented on src/main/resources/static/js/users/signup.js:
```diff
@@ -0,0 +1,49 @@
+            const passwordCheck = this.passwordCheckField.value;
+            const name = this.nameField.value;
+
+            console.log(name);
```
console.log는 빼도 괜찮을거같슴니다~

CHANMYUNG commented on src/main/java/com/woowahan/techcamp/recipehub/user/controller/UserRestController.java:
```diff
@@ -19,6 +18,12 @@
+    @PostMapping("/signup")
+    @ResponseStatus(HttpStatus.CREATED)
+    public void signup(@Valid @RequestBody SignupDto dto) {
+        userService.add(dto);
```
add보다는 create나 있는 그대로 signup이 어떨까요?

CHANMYUNG reviewed (changes requested):
구현 잘 했네요.

CHANMYUNG commented on src/main/resources/templates/users/login.hbs:
```diff
@@ -30,14 +30,15 @@
         </div>
     </div>
 </section>
+<script src="/js/users/login.js"></script>
```
수고하셨어요 ㅠㅠ 죄송합니다 ㅋㅋ

ryuneeee commented on src/main/java/com/woowahan/techcamp/recipehub/user/controller/UserRestController.java:
```diff
@@ -19,6 +18,12 @@
 
+    @PostMapping("/signup")
+    @ResponseStatus(HttpStatus.CREATED)
+    public void signup(@Valid @RequestBody SignupDto dto) {
```
Return이 빠진것 같아요~ 😢 이걸 void로 해야 맞을까요?

minzzzy replied:
그렇군요. 나중에 response model이 생기면 생성된 user에 대한 정보를 응답으로 보내는 것이 좋을까요?

ryuneeee replied:
Cool 🏝

ryuneeee commented on src/main/java/com/woowahan/techcamp/recipehub/user/service/UserService.java:
```diff
@@ -1,22 +1,30 @@
+    public User add(SignupDto dto) {
+        if (userRepository.existsByEmail(dto.getEmail())) {
+            throw new ConflictException();
+        }
```
이거 한 줄로 줄일수도 있을 것 같아요!
특정 유저가 존재하는지를 알려면 DB에 어떤 Query를 날려야 할까요? 
또, JPA는 existsByEmail에 대해서 어떤 Query를 만들어낼까요?

내일 같이 고민해봐요~

ryuneeee commented on src/test/java/com/woowahan/techcamp/recipehub/user/service/UserServiceTest.java:
```diff
@@ -22,14 +28,42 @@
 
     @Mock
-    private UserRepository userRepository;
+    private PasswordEncoder passwordEncoder;
```
userRepository와 passwordEncoder가 서로 크...크로스가 되었습니다! 멋져요👍

ryuneeee commented on src/main/resources/static/js/users/signup.js:
```diff
@@ -0,0 +1,49 @@
+                    passwordCheck,
+                    name
+                }),
+                onSuccess: () => {
```
ArrowFunction 적극사용 좋아요좋아요🙏

ryuneeee commented on src/main/resources/static/js/users/signup.js:
```diff
@@ -0,0 +1,49 @@
+
+            console.log(name);
+
+            fetchManager({
```
장기적으로 생각해 본다면 registerEvents()가 fetchManager까지 관심을 가져야할까요?
지금은 코드가 길지 않으니 괜찮지만, 한 번쯤 생각해보기~ 
(디테일한 코드를 보지 않고 어떤 이벤트를 등록하는지 어떻게 의도를 드러낼지도 고민해봅시당🧐)

ryuneeee commented on src/test/java/com/woowahan/techcamp/recipehub/user/dto/SignupDtoTest.java:
```diff
@@ -0,0 +1,95 @@
+
+    @Test
+    public void valid() {
+        assertConstraintViolations(signupDtoBuilder.build(), 0);
```
기본적인 사용 테스트 굿✌️ 이러면 별도의 문서가 필요 없죠

ryuneeee commented on src/main/java/com/woowahan/techcamp/recipehub/user/controller/UserController.java:
```diff
@@ -3,16 +3,26 @@
     public String userLogin(HttpSession session) {
         if (SessionUtils.isLoggedIn(session)) {
-            return "index";
+            return "redirect:/";
```
최고 👻 👍 👀🧠

ryuneeee commented on src/main/java/com/woowahan/techcamp/recipehub/user/dto/SignupDto.java:
```diff
@@ -0,0 +1,53 @@
+    @Size(min = 2, max = 40)
+    private String name;
+
+    @AssertTrue(message = "")
```
1. Validation을 위한 @AssertTrue Annotation👍 줍줍합니다
이거 찾아봤는데 저도 잘 감이 안와서, 내일 설명 부탁드릴게요🙏
2. matchPassword는 다른데서 쓰지 않는것 같아요. (사용한다면, 테스트 케이스 만들어주시면 감사!)

ryuneeee reviewed (approved):
좋아요. 전 Approve!

minzzzy committed:
refactor: Reflect feedback

- Change Boolean object to primitive type
- Rename method from 'add' to 'create'
- Remove unused message

minzzzy committed:
refactor: Return user when sign up

Lahavas reviewed (approved):
승인합니다~

Lahavas committed:
Merge branch 'develop' into feature/signup
